chore!: unify sources configuration handling (#560)
open-feature/flagd: docs/configuration/configuration.md

| provider    | required `string`  | Provider type - `file`, `kubernetes`, `http` or `grpc`                                                                                       |
| bearerToken | optional `string`  | Used for http sync and token get appended to `Authorization` header with [bearer schema](https://www.rfc-editor.org/rfc/rfc6750#section-2.1) |
| providerID  | optional `string`  | Value binds to grpc connection's providerID field. GRPC server implementations may use this to identify connecting flagd instance            |
| selector    | optional `string`  | Value binds to grpc connection's selector field. GRPC server implementations may use this to filter flag configurations                      |
| grpcSecure  | optional `boolean` | Used to enable secure TLS connectivity for grpc sync. Default(ex:- if unset) is false, which will use insecure grpc connection               |
| certPath    | optional `string`  | Used for grpcs sync when TLS certificate is needed. If not provided, system certificates will be used for TLS connection                     |

> The `uri` field values **do not** need to follow the [URI patterns](#uri-patterns), the provider type is instead derived

Reviewer: the phrasing here could be stronger, 'do not need' feels as though it would still work, whereas flagd would error if you do

Author: Thanks, yes will update this :)

Author: Updated the documentation to emphasize the difference between uri patterns and source configuration URI

# Configuration

> <!-- TOC -->

Author: Used idea md plugin [1]. I am not super familiar with this tool so I will create an issue out of this suggestion :) 

[1] - https://www.jetbrains.com/help/idea/markdown.html

> <!-- TOC -->

Author: Added a dedicated issue to improve this - https://github.com/open-feature/flagd/issues/570

* [Configuration](#configuration)
  * [Sync providers](#sync-providers)
    * [Kubernetes provider](#kubernetes-provider)
    * [Filepath provider](#filepath-provider)
    * [Remote provider](#remote-provider)
    * [GRPC provider](#grpc-provider)
  * [Sync provider configurations](#sync-provider-configurations)
    * [URI patterns](#uri-patterns)
    * [Source Configuration](#source-configuration)
<!-- TOC -->

`flagd` supports configuration via config file, environment variables and start-up flags. In cases of a conflict,
start-up flags have the highest priority, followed by environment variables and config file.

Supported start-up flags are documented (auto-generated) [here](./flagd_start.md).

Environment variable keys are uppercase, prefixed with `FLAGD_` and all `-` are replaced with `_`. For example,
`metrics-port` flag in environment variable form is `FLAGD_METRICS_PORT`.

Config file expects the keys to have the exact naming as startup-flags flags.

## Sync providers

> Sync providers are a core part of flagd. They are the sources of feature flag configurations for evaluations. flagd
> currently support the following sync providers,

Reviewer 2: ```suggestion
Sync providers are a core part of flagd; they are the abstraction that enables different sources for feature flag configurations.
flagd currently support the following sync providers:
```

* [Kubernetes provider](#kubernetes-provider)
* [Filepath Configuration](#filepath-provider)
* [Remote Configuration](#remote-provider)
* [GRPC Configuration](#grpc-provider)

### Kubernetes provider

> The Kubernetes sync provider allows flagD to connect to a Kubernetes cluster and evaluate flags against a specified

Reviewer 2: ```suggestion
The Kubernetes sync provider allows flagd to connect to a Kubernetes cluster and evaluate flags against a specified
```

FeatureFlagConfiguration resource as defined within
the [open-feature-operator](https://github.com/open-feature/open-feature-operator/blob/main/apis/core/v1alpha1/featureflagconfiguration_types.go)
spec.

To use an existing FeatureFlagConfiguration custom resource, start flagD with the following command:

```shell
flagd start --uri core.openfeature.dev/default/my_example
```

In this example, `default/my_example` expected to be a valid FeatureFlagConfiguration resource, where `default` is the
namespace and `my_example` being the resource name.

> ### Filepath provider

Reviewer 2: Thanks for adding blurbs about each provider, these are great!

> File path sync provider reads and watch the source file for updates(ex:- changes and deletions).

Reviewer 2: ```suggestion
The file path sync provider reads and watch the source file for updates(ex:- changes and deletions).
```

```shell
flagd start --uri file:etc/featureflags.json
```

In this example, `etc/featureflags.json` is a valid feature flag configuration file accessible by the flagd runtime.

### Remote provider

> Remote sync provider fetch flags from a remote source and periodically poll the source for flag configuration updates.

Reviewer 2: ```suggestion
The HTTP sync provider fetch flags from a remote source and periodically poll the source for flag configuration updates.
```

We've generally tried to replace "remote" to "http". "grpc" is also possibly "remote" so "http" is a better name.

```shell
flagd start --uri https://my-flag-source.json
```

In this example, `https://my-flag-source.json` is a remote endpoint responding valid feature flag configurations when
invoked with **HTTP GET** request.

### GRPC provider

> GRPC sync provider stream flag configurations from a grpc sync provider implementation. This stream connection is ruled

Reviewer 2: ```suggestion
The GRPC sync provider stream flag configurations from a grpc sync provider implementation. This stream connection is ruled
```

|-------------|--------------------|----------------------------------------------------------------------------------------------------------------------------------------------|
| uri         | required `string`  | Flag configuration source of the provider                                                                                                    |
| provider    | required `string`  | Provider type - `file`, `kubernetes`, `http` or `grpc`                                                                                       |
| bearerToken | optional `string`  | Used for http sync and token get appended to `Authorization` header with [bearer schema](https://www.rfc-editor.org/rfc/rfc6750#section-2.1) |
| providerID  | optional `string`  | Value binds to grpc connection's providerID field. GRPC server implementations may use this to identify connecting flagd instance            |
| selector    | optional `string`  | Value binds to grpc connection's selector field. GRPC server implementations may use this to filter flag configurations                      |

> | grpcSecure  | optional `boolean` | Used to enable secure TLS connectivity for grpc sync. Default(ex:- if unset) is false, which will use insecure grpc connection               |

Reviewer 2: Is there a reason you chose `grpcSecure` instead of something like `tls` which is more generic and might even apply to yet-to-be-added syncs?

Reviewer 2: I don't think it needs to be done in this PR, but the HTTP sync might be improved if it was made more consistent with GRPC. It should also be able to use `tls` and `certpath` (I think). IMHO that's a reason to use `tls` right now. `certPath` is sot of the same - it's grpc specific right now but the same concept applies to HTTP.

Reviewer 2: Agred

Author: Updated to use `tls` [1]

[1] - https://github.com/open-feature/flagd/blob/2a4783ef61eb3756e7a3f99aae596392906cd765/docs/configuration/configuration.md#source-configuration

Sync providers,

* `file` - config/samples/example_flags.json
* `http` - <http://my-flag-source.json/>
* `kubernetes` - default/my-flag-config
* `grpc`(insecure) - grpc-source:8080
* `grpc`(secure) - my-flag-source:8080

Startup command,

```sh
./bin/flagd start 
--sources='[{"uri":"config/samples/example_flags.json","provider":"file"},
            {"uri":"http://my-flag-source.json","provider":"http","bearerToken":"bearer-dji34ld2l"},
            {"uri":"default/my-flag-config","provider":"kubernetes"},
            {"uri":"grpc-source:8080","provider":"grpc"},
            {"uri":"my-flag-source:8080","provider":"grpc", "certPath": "/certs/ca.cert", "grpcSecure": true, "providerID": "flagd-weatherapp-sidecar", "selector": "source=database,app=weatherapp"}]'
```

Configuration file,

```yaml
sources:
  - uri: config/samples/example_flags.json
    provider: file
  - uri: http://my-flag-source.json
    provider: http
    bearerToken: bearer-dji34ld2l
  - uri: default/my-flag-config
    provider: kubernetes
  - uri: my-flag-source:8080
    provider: grpc
  - uri: my-flag-source:8080
    provider: grpc
    certPath: /certs/ca.cert

>     grpcSecure: true

Author: > What happens if `certPath` is set but `secure` is not

Right now, this results in an `insecure` connection [1] . I am not against making this an error. Other option is to log a warning and continue :) 

[1] - https://github.com/open-feature/flagd/blob/8501dc0648e6c48954d871cf62f46016aa4cc490/core/pkg/sync/grpc/credentials/builder.go#L24

>     grpcSecure: true

Reviewer 3: IMHO we should not fail, since the connection might still work. I am also in favor of a warning log.

>     grpcSecure: true

Author: @toddbaert @thisthat I went ahead with an error. The justification here is to fail fast.

If cert path is provided for an insecure (`tls:false`) sync, then this is a wrong configuration. The check is done at grpc credentials module [1] as this is the only place we currently use this config option. 

[1] - https://github.com/open-feature/flagd/blob/2a4783ef61eb3756e7a3f99aae596392906cd765/core/pkg/sync/grpc/credentials/builder.go#L24-L28

>     grpcSecure: true

Reviewer 2: Works for me!
